File: read_fast.py
#!/usr/bin/env python3
"""Fast encoder read test — measures actual achievable poll rate.

Reads as fast as possible for a few seconds, prints a summary
line every 500ms, and reports the achieved sample rate.

Usage:
    python3 read_fast.py
    python3 read_fast.py --duration 10
"""

import argparse
import time
import numpy as np
from quanser.hardware import HIL

CPR = 2048
COUNTS_TO_DEG = 360.0 / CPR
COUNTS_TO_RAD = 2.0 * np.pi / CPR

PEND_OFFSET = -1026  # from calibration (upright)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--duration", type=float, default=5.0)
    args = parser.parse_args()

    card = HIL("qube_servo3_usb", "0")
    enc_ch = np.array([0, 1], dtype=np.uint32)
    buf = np.zeros(2, dtype=np.int32)

    card.set_encoder_counts(enc_ch, 2, np.array([0, 0], dtype=np.int32))

    total_samples = 0
    t_start = time.perf_counter()
    last_print = t_start
    print_interval = 0.5  # print every 500ms

    print(f"Reading for {args.duration}s as fast as possible...")
    print(f"{'elapsed':>8s}  {'rate':>8s}  {'motor':>9s}  {'pendulum':>9s}  {'pend(cal)':>9s}")
    print("-" * 55)

    try:
        while True:
            card.read_encoder(enc_ch, 2, buf)
            total_samples += 1

            now = time.perf_counter()
            elapsed = now - t_start

            if elapsed >= args.duration:
                break

            if now - last_print >= print_interval:
                rate = total_samples / elapsed
                motor_deg = buf[0] * COUNTS_TO_DEG
                pend_raw_deg = buf[1] * COUNTS_TO_DEG
                pend_cal_deg = (buf[1] - PEND_OFFSET) * COUNTS_TO_DEG
                print(f"{elapsed:7.2f}s  {rate:7.0f}Hz  {motor_deg:8.2f}°  {pend_raw_deg:8.2f}°  {pend_cal_deg:8.2f}°")
                last_print = now

    except KeyboardInterrupt:
        pass
    finally:
        elapsed = time.perf_counter() - t_start
        rate = total_samples / elapsed
        card.close()

        print("-" * 55)
        print(f"Total samples: {total_samples}")
        print(f"Elapsed:       {elapsed:.3f}s")
        print(f"Average rate:  {rate:.0f} Hz")
        print(f"Period:        {1e6/rate:.1f} µs/sample")


if __name__ == "__main__":
    main()

File: cpp/read_fast.cpp
// Fast encoder read test — measures achievable poll rate.
// Build:  g++ -O2 -std=c++17 -o read_fast read_fast.cpp -lhil -lquanser_common -lm
// Run:    ./read_fast

#include <hil.h>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <ctime>

static constexpr int    COUNTS_PER_REV = 2048;
static constexpr double COUNTS_TO_DEG  = 360.0 / COUNTS_PER_REV;
// No count offset. Encoder zero = hanging down, 180° = upright.
// "pend(cal)" shows angle from upright: 0° = upright, ±180° = down.
static constexpr double PEND_OFFSET_DEG = 180.0;

static double now_sec() {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec + ts.tv_nsec * 1e-9;
}

int main(int argc, char** argv) {
    double duration = 5.0;
    for (int i = 1; i < argc; i++)
        if (!strcmp(argv[i], "--duration") && i + 1 < argc) duration = atof(argv[++i]);

    t_card card;
    t_error result = hil_open("qube_servo3_usb", "0", &card);
    if (result < 0) {
        fprintf(stderr, "Failed to open (error %d)\n", result);
        return 1;
    }

    const t_uint32 enc_ch[] = {0, 1};
    t_int32 buf[2];
    t_int32 zero[] = {0, 0};
    hil_set_encoder_counts(card, enc_ch, 2, zero);

    double t_start  = now_sec();
    double t_print  = t_start;
    long   samples  = 0;

    printf("Reading for %.0fs as fast as possible...\n", duration);
    printf("%8s  %8s  %9s  %9s  %9s\n", "elapsed", "rate", "motor", "pendulum", "pend(cal)");
    printf("------------------------------------------------------\n");

    while (true) {
        hil_read_encoder(card, enc_ch, 2, buf);
        samples++;

        double now = now_sec();
        double elapsed = now - t_start;

        if (elapsed >= duration) break;

        if (now - t_print >= 0.5) {
            double rate      = samples / elapsed;
            double motor_deg = buf[0] * COUNTS_TO_DEG;
            double pend_deg  = buf[1] * COUNTS_TO_DEG;
            double pend_cal  = buf[1] * COUNTS_TO_DEG - PEND_OFFSET_DEG;
            printf("%7.2fs  %7.0fHz  %8.2f°  %8.2f°  %8.2f°\n",
                   elapsed, rate, motor_deg, pend_deg, pend_cal);
            t_print = now;
        }
    }

    double elapsed = now_sec() - t_start;
    double rate = samples / elapsed;
    hil_close(card);

    printf("------------------------------------------------------\n");
    printf("Total samples: %ld\n", samples);
    printf("Elapsed:       %.3fs\n", elapsed);
    printf("Average rate:  %.0f Hz\n", rate);
    printf("Period:        %.1f us/sample\n", 1e6 / rate);

    return 0;
}
